### ISSUE TITLE
[NNC] Vectorization caused wrong results

### ISSUE THREAD

**Issue by huiguoo** [pytorch/pytorch#56741]
## 🐛 Bug

After performing NNC's vectorization on the innermost loop of a loopnest, the produced result become incorrect.

 
## To Reproduce

Code to perform vectorization and print before/after stmts:
```
simplified_stmts = loopnest.simplify().stmts()
stmt = simplified_stmts[-1]
print(f"Original stmt:\n{stmt}")
print(f"Before vectorization:\n{loopnest.root_stmt()}")
stmt = loopnest.flatten([stmt])
stmt = loopnest.normalize(stmt)
stmt = loopnest.get_innermost_loops_for(Pconv.data())[-1]
loopnest.vectorize(stmt)
print(f"After vectorization:\n{loopnest.root_stmt()}") 
```

Original stmt:
```
for (int g = 0; g < 96; g++) {
  for (int w = 1; w < 13; w++) {
    for (int r = 0; r < 5; r++) {
      for (int s = 0; s < 5; s++) {
        for (int h = 1; h < 13; h++) {
          conv[0, g, h, w] = (conv[0, g, h, w]) + (image[0, g, (2 * h + r) - 2, (2 * w + s) - 2]) * (weight[g, 0, r, s]);
        }
      }
    }
  }
}
```
Before vectorization
```
...
  for (int g_4 = 0; g_4 < 96; g_4++) {
    for (int w_2 = 1; w_2 < 13; w_2++) {
      for (int r_4 = 0; r_4 < 5; r_4++) {
        for (int s_4 = 0; s_4 < 5; s_4++) {
          for (int h_2 = 1; h_2 < 13; h_2++) {
            conv[(((0 + 0 * (((1 * 14) * 14) * 96)) + g_4 * ((1 * 14) * 14)) + h_2 * (1 * 14)) + w_2 * 1] = (conv[(((0 + 0 * (((1 * 14) * 14) * 96)) + g_4 * ((1 * 14) * 14)) + h_2 * (1 * 14)) + w_2 * 1]) + (image[(((0 + 0 * (((1 * 28) * 28) * 96)) + g_4 * ((1 * 28) * 28)) + ((2 * h_2 + r_4) - 2) * (1 * 28)) + ((2 * w_2 + s_4) - 2) * 1]) * (weight[(((0 + g_4 * (((1 * 5) * 5) * 1)) + 0 * ((1 * 5) * 5)) + r_4 * (1 * 5)) + s_4 * 1]);
          }
        }
      }
    }
  }

```
After vectorization
```
...
  for (int g_4 = 0; g_4 < 96; g_4++) {
    for (int w_2 = 1; w_2 < 13; w_2++) {
      for (int r_4 = 0; r_4 < 5; r_4++) {
        for (int s_4 = 0; s_4 < 5; s_4++) {
{
            conv[Ramp((w_2 + 196 * g_4) + 14, 14, 13)] = (conv[Ramp((w_2 + 196 * g_4) + 14, 14, 13)]) + (image[Ramp((((s_4 + 2 * w_2) + 784 * g_4) + 28 * r_4) - 2, 56, 13)]) * (Broadcast(weight[(25 * g_4 + s_4) + 5 * r_4], 13));
          }        }
      }
    }
  }

```

Comparing to the correct results:
```
With rtol=0.0001 and atol=1e-05, found 1152 element(s) (out of 18816) whose difference(s) exceeded the margin of error (including 0 nan comparisons). The greatest difference was 1.9056527024507724e+26 (1.9056527024507724e+26 vs. -0.2777479588985443), which occurred at index (0, 95, 13, 9).
```
## Expected behavior

Produce the same results as without vectorization

**Comment by bertmaher** [comment]
Interesting.  Does this happen even if the H dimension is 16 (or some other multiple of the actual machine vector width)?

**Comment by huiguoo** [comment]
Probably no? I remember tried vectorization for a `h=8` case and that one works well. So is this one of the requirements to perform vectorization? - I expect it to vectorize 8 iterations and put the left in a tail loop?

**Comment by bertmaher** [comment]
Really it *should* work with arbitrary width, I think.  I'm not sure what's going wrong with size 13.  It could be an llvm bug, since I suspect vectorize by 13 isn't used often.

I think ideally we'd generate masked code with width 16, or maybe tail loops.

**Comment by ZolotukhinM** [comment]
The problem is not even that the size is awkward, the problem is that the size should be 13-1=12 since the loop starts from 1.

I wrote a small repro for this:
```
import torch

class kernel_arena_scope(object):
    def __enter__(self):
        self.scope = torch._C._te.KernelScope()

    def __exit__(self, typ, val, traceback):
        self.scope = None

def repro(vectorize):
    print ("********************************")
    print ("Vectorize = ", vectorize)
    with kernel_arena_scope():
        dtype = torch._C._te.Dtype.Float
        N = 13
        M = 2

        dn = torch._C._te.ExprHandle.int(N)
        dm = torch._C._te.ExprHandle.int(M)

        X = torch._C._te.Placeholder('X', dtype, [dm, dn])
        Y = torch._C._te.Placeholder('Y', dtype, [dm, dn])

        dim_args = [torch._C._te.DimArg(*args) for args in [(dm, 'm'), (dn, 'n')]]

        def compute(m, n):
            return X.load([m, n]) + Y.load([m, n])
        Z = torch._C._te.Compute('Z', dim_args, compute)
        loopnest = torch._C._te.LoopNest([Z])
        print("original", loopnest)
        loops = loopnest.get_loops_for(Z)
        loopnest.slice_head(loops[1], 1)
        loopnest.simplify()
        print("after peel", loopnest)
        loops = loopnest.get_innermost_loops_for(Z.buf())
        if vectorize:
            loopnest.vectorize(loops[1])
            loopnest.simplify()
            print("after vectorize", loopnest)
        else:
            print("Skip vectorization")

        tX = torch.arange(0, N * 3, dtype=torch.float).reshape(3, N) + 1.
        tY = torch.arange(0, N * 3, dtype=torch.float).reshape(3, N) + 2.
        tZ = torch.empty(2, N)

        codegen = torch._C._te.construct_codegen('ir_eval', loopnest.root_stmt(), [torch._C._te.BufferArg(x) for x in [X, Y, Z]])
        codegen.call([tX, tY, tZ])
        print(tZ)
        assert(torch.allclose(tZ, (tX + tY)[0:2, 0:N]))
        print("Success!")

repro(vectorize=False) # Succeeds
repro(vectorize=True)  # Fails
```
Output:
```
********************************
Vectorize =  False
original {
  for (int m = 0; m < 2; m++) {
    for (int n = 0; n < 13; n++) {
      Z[m, n] = (X[m, n]) + (Y[m, n]);
    }
  }
}
after peel {
  for (int m = 0; m < 2; m++) {
    Z[m, 0] = (X[m, 0]) + (Y[m, 0]);
    for (int n = 1; n < 13; n++) {
      Z[m, n] = (X[m, n]) + (Y[m, n]);
    }
  }
}
Skip vectorization
tensor([[ 3.,  5.,  7.,  9., 11., 13., 15., 17., 19., 21., 23., 25., 27.],
        [29., 31., 33., 35., 37., 39., 41., 43., 45., 47., 49., 51., 53.]])
Success!
********************************
Vectorize =  True
original {
  for (int m = 0; m < 2; m++) {
    for (int n = 0; n < 13; n++) {
      Z[m, n] = (X[m, n]) + (Y[m, n]);
    }
  }
}
after peel {
  for (int m = 0; m < 2; m++) {
    Z[m, 0] = (X[m, 0]) + (Y[m, 0]);
    for (int n = 1; n < 13; n++) {
      Z[m, n] = (X[m, n]) + (Y[m, n]);
    }
  }
}
after vectorize {
  for (int m = 0; m < 2; m++) {
    Z[m, 0] = (X[m, 0]) + (Y[m, 0]);
    Z[Ramp(13 * m + 1, 1, 13)] = (X[Ramp(13 * m + 1, 1, 13)]) + (Y[Ramp(13 * m + 1, 1, 13)]);
  }
}
double free or corruption (out)
```

**Comment by navahgar** [comment]
@ZolotukhinM So, should we just enforced that the loop is normalized before vectorization?

**Comment by ZolotukhinM** [comment]
That's one way to fix it, yes. I think @huiguoo mentioned that there have been some problems with normalization too?

**Comment by navahgar** [comment]
Fixed by #59423
